contracts/sui/xcall/sources/cluster_connection/cluster_state.move

```diff
@@ -0,0 +1,356 @@
+            conn_sn: 0,
+            receipts: vec_map::empty(),
+            balance:balance::zero(),
+            validators: vector::empty(),
+            validators_threshold:0
+        }
+    }
+
+    fun get_validator(pub_key:String):Validator{
+        let (pubkey,sui_address)=pubkey_to_sui_address(&pub_key);
+        Validator{
+            pub_key:pubkey,
+            sui_address:sui_address
+        }
+    }
+
+    public(package) fun get_next_conn_sn(self:&mut State):u128 {
+        let sn=self.conn_sn+1;
+        self.conn_sn=sn;
+        sn
+    }
+
+    public fun get_fee(self: &State, netId: &String, need_response: bool): u64 {
+        let fee: u64 = if (need_response == true) {
+            utils::get_or_default(&self.message_fee, netId, 0)
+                + utils::get_or_default(&self.response_fee, netId, 0)
+        } else {
+            utils::get_or_default(&self.message_fee, netId, 0)
+        };
+        fee
+    }
+
+    public(package) fun set_fee(self: &mut State, net_id: String, message_fee: u64, response_fee: u64,caller:address) {
+        if (vec_map::contains(&self.message_fee,&net_id)){
+            vec_map::remove(&mut self.message_fee,&net_id);
+        };
+        if (vec_map::contains(&self.response_fee,&net_id)){
+            vec_map::remove(&mut self.response_fee,&net_id);
+        };
+        vec_map::insert(&mut self.message_fee, net_id, message_fee);
+        vec_map::insert(&mut self.response_fee, net_id, response_fee);
+    }
+
+    public(package) fun check_save_receipt(self: &mut State, net_id: String, sn: u128) {
+        let receipt_key = ReceiptKey { nid: net_id, conn_sn: sn };
+        assert!(!vec_map::contains(&self.receipts, &receipt_key), 100);
+        vec_map::insert(&mut self.receipts, receipt_key, true);
+    }
+
+    public(package) fun get_receipt(self: &State, net_id: String, sn: u128): bool {
+        let receipt_key = ReceiptKey { nid: net_id, conn_sn: sn };
+        vec_map::contains(&self.receipts, &receipt_key)
+    }
+
+    public(package) fun deposit(self:&mut State,balance:Balance<SUI>){
+        balance::join(&mut self.balance,balance);
+
+    }
+    
+    public(package) fun claim_fees(self:&mut State,ctx:&mut TxContext){
+        let total= self.balance.withdraw_all();
+        let coin= coin::from_balance(total,ctx);
+        transfer::public_transfer(coin,ctx.sender());
+
+    }
+
+    public(package) fun verify_signatures(self:&State,msg:vector<u8>,signatures:vector<vector<u8>>){
+        let threshold=self.get_validator_threshold();
+        let validators=self.get_validators().map!(|validator| validator.pub_key);
+        assert!(signatures.length() >= threshold, NotEnoughSignatures);
+        let mut i = 0;
+        let mut unique_verified_pubkey = vector::empty();
+        while (i < signatures.length()) {
+            let signature = signatures.borrow(i);
+            let pub_key = get_pubkey_from_signature(signature);
```

Comment:
are we sure that we will receive raw_signature(scheme+pubkey+signature) here ? plz make sure with relay guys, in my opinion they may only send signature here.

Reply:
this one was discussed with Biru daii, and planned this!